[IA-2389][IA-2596] Custom hardware config for Galaxy apps and other small fixes

kyuksel opened the pull request:
[JIRA ticket
](https://broadworkbench.atlassian.net/browse/IA-2389)

### Highlights
- Galaxy users can now specify how many VMs to run their workflows on and how many/much vCPUs/memory each VM will have. 
- They can also select a size for their persistent data disk (metadata disk will remain fixed size of `10 GB`, which isn't directly exposed to users).
- Allowing users to update their hardware configuration is out of scope. To make this more clear, added a note to the info page right before app creation. Furthermore per UX request, revamped that page's visuals with Font Awesome icons for each bullet.
- Tried to make the app config customization window's layout as consistent as possible with that of the runtime modal's.
- Fixed a bug where hourly cost was being displayed for Galaxy disks although the text said `per month` 🤦🏼 
- Added a tooltip for the `Delete` button when a Galaxy app is in states like `pausing`, `paused` and `resuming`, informing user that they resume first. Also disabled the `Pause` and `Delete` buttons where applicable.
- Filtered out some of the low cpu/memory combinations based on feedback from the Galaxy team regarding usage patterns.
- Stylistic changes to fix typos and standardize user-facing messaging across Runtime and Galaxy modals.

### Notes
- Added plumbing to route the list of Galaxy data (NFS) disks from `WorkspaceContainer` through `Notebooks` and `RuntimeManager` to `NewGalaxyModal`.
- Put `MachineSelector` outside `NewGalaxyModal` so it can be called as a component, consistently with `NewRuntimeModal`'s. The other way, it was causing a very strange behaviour on the `Nodes` `NumberInput` component where incrementing and inserting double-digit numbers weren't working properly.
- Changed wording of messages displayed to users regarding what the cloud environment is doing at a given stage, in hopes of making them more grammatically consistent and also using the terms `pause/resume` consistently instead of a mix of them and `stop/start`. There is another ticket in progress for doing that throughout the codebase.
- Fixed typos.
- Punted on the `TODO`s around refactoring since the UI redesign slated for Q2 will probably make them irrelevant but kept them in in case that's not the case.

### Verification
- Manually clicked through lifecycles of a Galaxy app.
- [galaxy_screenshots_210326.pdf](https://github.com/DataBiosphere/terra-ui/files/6212342/galaxy_screenshots_210326.pdf) shows a lifecycle for a Galaxy app showing some of the changes in this PR. Note that the final versions are slightly different upon PR and UX feedback.

kyuksel committed:
Galaxy cost breakdown

kyuksel committed:
Pluralize Preemptible

kyuksel committed:
Merge remote-tracking branch 'origin/dev' into ky_ia_2389_custom_nodepool_config

kyuksel committed:
styling

kyuksel committed:
darker background colour

kyuksel committed:
Application configuration

kyuksel committed:
Add sections 'Cloud compute profile' and 'Persistent disk size'

kyuksel committed:
fix some styling issues

kyuksel committed:
Fix DiskSelector inputs

kyuksel committed:
grid styling

kyuksel committed:
Fix disk size selection issues

kyuksel committed:
Merge remote-tracking branch 'origin/dev' into ky_ia_2389_custom_nodepool_config

kyuksel committed:
Fix CPU and Memory selectors. Minor style change.

kyuksel committed:
Add space above the horizontal line to match NewRuntimeModal

kyuksel committed:
Update app creation payload to customize nodepool config

kyuksel committed:
Plumbing to make Galaxy data disk info available to components

kyuksel committed:
Get Galaxy data disk info through RuntimeManager. useEffect for Galaxy modal

kyuksel committed:
Dynamic cost display for running apps. Clean-up.

kyuksel committed:
Fix and reformat messaging

kyuksel committed:
Merge remote-tracking branch 'origin/dev' into ky_ia_2389_custom_nodepool_config

kyuksel committed:
Avoid race condition at first render time by checking for dataDisk nullity

kyuksel committed:
Fix Nodes button issue. Update some wording.

kyuksel committed:
Move MachineSelector out of the modal

kyuksel committed:
Move some const's up for visibility

kyuksel commented on src/pages/workspaces/workspace/Notebooks.js:
```diff
@@ -307,15 +308,14 @@ const Notebooks = _.flow(
-      return app ?
+    const getGalaxyText = (app, galaxyDataDisks) => {
+      const dataDisk = currentDataDisk(app, galaxyDataDisks)
+      return (app && dataDisk) ?
```
Added `&& dataDisk` to prevent the error that occasionally happened (probably due to race condition at the first render of the component)

kyuksel commented on src/components/NewGalaxyModal.js:
```diff
@@ -250,3 +322,59 @@ export const NewGalaxyModal = _.flow(
+            isClearable: false,
+            onlyInteger: true,
+            value: value.numNodes,
+            onChange: n => onChange(prevState => { return { ...prevState, numNodes: n } })
```
Found out that updating JS objects with multiple (nested or not) fields works differently for function components than class components. For the latter, if only one field of such a JS object is updated, it gets merged as desired. For the former, one has to update the entire object, which is why I used the object spread syntax with `…` as suggested by [this blog](https://blog.logrocket.com/a-guide-to-usestate-in-react-ecb9952e406c).

kyuksel committed:
Updated machine types offered and changed PD default size based on Galaxy team feedback

ehanna4 commented on src/components/NewGalaxyModal.js:
```diff
@@ -186,59 +198,118 @@ export const NewGalaxyModal = _.flow(
 
     return !app ?
-      isTitle ? 'Cloud environment' : 'Environment will consist of an application and cloud compute.' :
+      'consists of application configuration, cloud compute and persistent disk(s).' :
```
Was this change recommended by UX? Mostly curious because in the cloud env version it says `"Cloud environments consist of an application configuration, cloud compute and a persistent disk" `so this seems like a change that makes it less similar rather than more similar

xuscrbw replied:
Going off of this, I will say that preference wise the two separate headings across the same sentence feels a bit off. I don't think there's actually any functional difference here, but was more of stylistic preference for me

kyuksel replied:
It wasn't particularly recommended by UX. I noticed that on the current runtime version below, `Cloud environment` is duplicated, and I wanted to remove duplication. I also wanted to be able to use text consistently across runtimes and apps, and since Galaxy apps have multiple disks, I turned `disk` into `disk(s)`. Note that both runtime and Galaxy modals now use it so it'll be the same. 

No functional difference indeed.

Do these address your questions?

![image](https://user-images.githubusercontent.com/5438223/112533684-98927f00-8d80-11eb-90be-e8f62d6d5142.png)

rtitle replied:
No strong opinions but perhaps `Consists` should be capitalized?

ehanna4 replied:
I can buy your rationale haha, though I agree with Rob I think capitalization would feel better. The bolded "Cloud Environment" is a title so it's feels funny to have it be like part of a fragmented sentence.

kyuksel replied:
Changed it to

-------------------------

### Cloud Environment
A cloud environment consists of application configuration, cloud compute and persistent disk(s).

-------------------------

as having a full sentence looked better to me. Let me know if any objections.

ehanna4 commented on src/components/NewGalaxyModal.js:
```diff
@@ -186,59 +198,118 @@ export const NewGalaxyModal = _.flow(
+    return div({ style: { ...styles.whiteBoxContainer, marginTop: '1rem' } }, [
+      div({ style: styles.headerText }, ['Persistent disk']),
+      div({ style: { marginTop: '0.5rem' } }, [
+        'Stores your analysis data '
```
```suggestion
        'Stores your analysis data.'
```

ehanna4 replied:
Wondering if having the documentation about persistent disks linked here (like it is in the cloud env version) would be beneficial? ie:
`Stores your analysis data. Learn more about Persistent disks and where your disk is mounted`

edit: oops just saw your TODO on this haha

xuscrbw commented:
<img width="287" alt="Screen Shot 2021-03-25 at 2 51 56 PM" src="https://user-images.githubusercontent.com/62293969/112528056-10a97680-8d7a-11eb-87ed-200890eafaad.png">
Noticing this:

1. I don't know if the cost here makes sense. Am using the default App for this. Also, The text seems to fall out of the box. Probably has to do with the cost per hr being longer. Not sure of a clean fix for this. @ehanna4 Have you run into this before?

kyuksel committed:
Add comment

kyuksel committed:
Merge remote-tracking branch 'origin/dev' into ky_ia_2389_custom_nodepool_config

xuscrbw reviewed (commented):
<img width="688" alt="Screen Shot 2021-03-25 at 3 12 58 PM" src="https://user-images.githubusercontent.com/62293969/112530307-acd47d00-8d7c-11eb-9f6e-ed2aaee97fc8.png">
Question about this view: right now I have an app running and it looks like the numbers here can still be changed. Is this how we are intending for this view to look? Not sure if changing these will be applied automatically, or if something else needs to be happening.

kyuksel commented:
@xuscrbw Great catch on the wrong cost display! There used to be a bug in persistent disk reporting, which I fixed but that introduced the one you found. 

> I don't know if the cost here makes sense.
I don't think we need to display the cost there but let me see if I can get some UX input as I may be missing some historical context.

kyuksel committed:
Improve the Galaxy cost display on the Notebooks page

kyuksel commented:
> Question about this view: right now I have an app running and it looks like the numbers here can still be changed. Is this how we are intending for this view to look?

@xuscrbw I wondered the same myself and then realized that the runtime modal has it the same way and decided to keep Galaxy's consistent with the runtime one. One difference is that Galaxy apps don't yet support updates for running instances so a user can play with configs (perhaps to see how that can change cost estimates) but nothing will change resource-wise.

kyuksel committed:
PR Feedback: 1)Fix total app cost 2)Add . to end of sentence and make runtime one's consistent

kyuksel committed:
UX feedback

kyuksel commented on src/components/NewGalaxyModal.js:
```diff
@@ -68,9 +82,11 @@ export const NewGalaxyModal = _.flow(
 
-  const deleteButton = h(ButtonSecondary, { style: { marginRight: 'auto' }, onClick: () => setViewMode('deleteWarn') }, ['Delete'])
-
   const renderActionButton = () => {
```
The changes in this function below were upon the UX request to add disabled `Delete` and `Pause` buttons to transitionary status modals.

ehanna4 commented on src/components/NewGalaxyModal.js:
```diff
@@ -83,23 +99,29 @@ export const NewGalaxyModal = _.flow(
+          ['ERROR', () => deleteButton],
+          [Utils.DEFAULT, () => {
+            return h(Fragment, [
+              deleteButton(true),
```
I think this should be `false`. Otherwise it is not allowing the user to delete the galaxy if it is errorred.

rtitle replied:
Seems the `ERROR` case is handled above on L118?

kyuksel replied:
Totally! Thanks for catching this 🙏🏼

kyuksel replied:
Oh actually, the `ERROR` case is above and does have a `deleteButton`. The one you pointed out is for undelete-able cases.

kyuksel replied:
(For some reason Rob's comment above hadn't shown up on my screen when I wrote the above replies  🤷🏼‍♂️ )

rtitle commented on src/components/NewGalaxyModal.js:
```diff
@@ -83,23 +99,29 @@ export const NewGalaxyModal = _.flow(
+            resumeButton()
           ])],
-          ['ERROR', () => deleteButton]
+          ['ERROR', () => deleteButton],
```
Should this be `deleteButton()` (do parens matter?)

kyuksel replied:
Good catch! Parentheses matter in this case because the `deleteButton()` function has a parameter with a default value, which doesn't seem to get set if no `()`.

kyuksel committed:
Don't disable Delete button when app is Error'ed

rtitle commented on src/components/NewGalaxyModal.js:
```diff
@@ -186,59 +207,118 @@ export const NewGalaxyModal = _.flow(
-        ['PRESTARTING', () => 'Cloud environment is preparing to start.'],
-        ['STARTING', () => `Cloud environment is starting. ${waitMessage}`],
-        ['RUNNING', () => 'Environment consists of an application and cloud compute.'],
+        ['STOPPED', () => `is paused.`],
```
I see `pause`, `stop`, `start`, and `resume` here. Can we use consistent language?

kyuksel replied:
Totally agreed - especially important for the user-facing ones. I've updated some things in my PR as I touched them but we have a [ticket in progress](https://broadworkbench.atlassian.net/browse/IA-2539) to go through the codebase for it.

xuscrbw replied:
For clarity, did you just mean the "is ..." text itself? Or do you mean, the statuses themselves?

kyuksel replied:
We can skip the statuses like `STOPPED` and `STARTING` above since they're not user-facing but there are some that _are_ user-facing like [this one in RuntimeManager](https://github.com/DataBiosphere/terra-ui/blob/dd5cd6d782bf878d3c33ad6ed0df05c673dd61f6/src/components/RuntimeManager.js#L310) which would be good to rename. What do you think?

xuscrbw replied:
I think this suggestion is good. Will look at the best way to accomplish this. Feels like we should have a helper function to make this work, might tackle this in its own PR.

rtitle commented on src/components/NewGalaxyModal.js:
```diff
@@ -83,23 +99,29 @@ export const NewGalaxyModal = _.flow(
+          ['ERROR', () => deleteButton],
+          [Utils.DEFAULT, () => {
+            return h(Fragment, [
+              deleteButton(),
```
I think it should be `deleteButton(true)` here -- otherwise it makes `STOPPING`, `STARTING` deletable which is not what we want. (Error is handled above).

rtitle replied:
Example from local testing. The app now is `Starting` and the `DELETE` button is clickable, which is not correct:
![image](https://user-images.githubusercontent.com/5368863/112678928-ef13c200-8e41-11eb-9cc1-d579d3af8e81.png)

rtitle commented on src/components/NewGalaxyModal.js:
```diff
@@ -186,59 +207,118 @@ export const NewGalaxyModal = _.flow(
+    return div({ style: { ...styles.whiteBoxContainer, marginTop: '1rem' } }, [
+      div({ style: styles.headerText }, ['Persistent disk']),
+      div({ style: { marginTop: '0.5rem' } }, [
+        'stores your analysis data. '
```
Just IMO, I feel like these text snippets should be capitalized and be complete sentences.

rtitle commented on src/components/NewGalaxyModal.js:
```diff
@@ -186,59 +207,118 @@ export const NewGalaxyModal = _.flow(
     return h(Fragment, [
       h(TitleBar, {
-        title: getEnvMessageBasedOnStatus(true),
+        title: 'Cloud environment',
```
Cloud Environment

rtitle commented on src/components/NewRuntimeModal.js:
```diff
@@ -821,7 +821,7 @@ export const NewRuntimeModal = withModalDrawer({ width: 675 })(class NewRuntimeM
             label({ htmlFor: id, style: styles.label }, ['Persistent disk size (GB)']),
             div({ style: { marginTop: '0.5rem' } }, [
-              'Stores your analysis data. ',
+              'stores your analysis data. ',
```
Same as above re capitalization

rtitle replied:
Can this be capital `S`?

kyuksel replied:
Made it a full sentence to be consistent with Galaxy modal's wording.

![Messages Image(696994228)](https://user-images.githubusercontent.com/5438223/113873844-2dcf4380-9783-11eb-823c-d6bf137233ff.jpeg)

rtitle commented on src/components/RuntimeManager.js:
```diff
@@ -341,7 +342,7 @@ export default class RuntimeManager extends PureComponent {
               div({ style: { fontSize: 12, fontWeight: 'bold' } }, 'Cloud Environment'),
               div({ style: { fontSize: 10 } }, [
-                span({ style: { textTransform: 'uppercase', fontWeight: 500 } },
+                span({ style: { textTransform: 'capitalize', fontWeight: 500 } },
```
Things look inconsistent for me now:
![image](https://user-images.githubusercontent.com/5368863/112678702-a22feb80-8e41-11eb-8206-f2cf85b3a3c0.png)

I think `uppercase` is fine..?

kyuksel replied:
+1 on the inconsistency. Fixed as in below:

![image](https://user-images.githubusercontent.com/5438223/113054744-c78d6400-9177-11eb-9c5f-c726dc1772bd.png)

Uppercasing signifies strong emphasis to me, at least subconsciously, which I doubt is what we want in this case.

rtitle replied:
Thanks.. am ok with `capitalize` instead of `uppercase`, as long as they're consistent. :)

kyuksel replied:
I'm all for consistency whenever possible 😉

rtitle commented:
Tested locally and it seems to be functional. Some comments above around presentation. In general I think the text snippets displayed in the various views should be capitalized and complete sentences. Curious what UX's opinion is though.

kyuksel committed:
Pull in dev branch changes and address PR feedback

ehanna4 commented on src/pages/workspaces/workspace/Notebooks.js:
```diff
@@ -363,7 +361,7 @@ const Notebooks = _.flow(
             onClick: () => setOpenGalaxyConfigDrawer(true)
           }, [
-            getGalaxyText()
+            getGalaxyText(app, galaxyDataDisks)
```
This tool tip is using `appIsSettingUp` so displays while app is creating but this button continues to say `create a galaxy app`. Probably should use that `appIsSettingUp` for this Galaxy text as well

kyuksel replied:
Thanks for this catch! I changed the logic around and it seems to work as intended ™

ehanna4 commented:
> > Question about this view: right now I have an app running and it looks like the numbers here can still be changed. Is this how we are intending for this view to look?
> 
> @xuscrbw I wondered the same myself and then realized that the runtime modal has it the same way and decided to keep Galaxy's consistent with the runtime one. One difference is that Galaxy apps don't yet support updates for running instances so a user can play with configs (perhaps to see how that can change cost estimates) but nothing will change resource-wise.

On this note - it feels weird to be able to play around with configs then hit the `launch` button and have nothing change. I wonder if we could make it clear that the user is just relaunching their existing app and the changes won't apply.

Maybe even on this view `Important: Please keep this tab open and logged in to Terra while using Galaxy.
Galaxy will open in a new tab.` could say like `Also please note to apply changes to your configuration please delete your existing app and create a new one with your changes blah blah` <- probably would need a UX input on that wording.

ehanna4 commented:
I don't see any update to mention PD in the delete view, when a user clicks delete is that killing everything or does the PD still exist? If it still exists how do we want users to be able to re-access it?

ehanna4 commented on src/components/NewGalaxyModal.js:
```diff
@@ -68,9 +82,11 @@ export const NewGalaxyModal = _.flow(
-  const deleteButton = h(ButtonSecondary, { style: { marginRight: 'auto' }, onClick: () => setViewMode('deleteWarn') }, ['Delete'])
-
   const renderActionButton = () => {
+    const deleteButton = (isDisabled = false) => h(ButtonSecondary, { disabled: isDisabled, style: { marginRight: 'auto' }, onClick: () => setViewMode('deleteWarn') }, ['Delete'])
```
nit: could be nice to include tooltips on these buttons when disabled

kyuksel committed:
PR feedback: Capitalize and use full sentences

kyuksel commented:
> I don't see any update to mention PD in the delete view, when a user clicks delete is that killing everything or does the PD still exist? If it still exists how do we want users to be able to re-access it?

@ehanna4 This PR doesn't change anything around persistence for Galaxy apps. Until we have the Galaxy Persistence ticket's changes in, we'll still be creating/deleting disks along with apps.

kyuksel committed:
Update => View since we don't support Galaxy app updating yet

kyuksel committed:
Merge remote-tracking branch 'origin/dev' into ky_ia_2389_custom_nodepool_config

kyuksel committed:
PR feedback: take into account appIsSettingUp()

kyuksel commented:
@ehanna4 

>On this note - it feels weird to be able to play around with configs then hit the launch button and have nothing change. I wonder if we could make it clear that the user is just relaunching their existing app and the changes won't apply. Maybe even on this view Important: Please keep this tab open and logged in to Terra while using Galaxy. Galaxy will open in a new tab. could say like Also please note to apply changes to your configuration please delete your existing app and create a new one with your changes blah blah <- probably would need a UX input on that wording. OR maybe near the launch button have an i or * with a note that it's the same app

Yea, good thinking! I'm thinking the heads-up page right before creating an app may also be a good place and it has more space to accommodate a wordy message.

![image](https://user-images.githubusercontent.com/5438223/113325827-e827fc00-92e6-11eb-95b9-7800a45b2cbf.png)

kyuksel committed:
PR feedback: Add heads-up about environment updates

kyuksel committed:
Reformat info page to be able to insert more than one icon

kyuksel committed:
Insert the desired icons

kyuksel committed:
use consistent font awesome

kyuksel committed:
Merge remote-tracking branch 'origin/dev' into ky_ia_2389_custom_nodepool_config

kyuksel commented:
@ehanna4 @rtitle @xuscrbw I believe I've addressed all of your comments as well as Joy's via [this diff](https://github.com/DataBiosphere/terra-ui/pull/2417/files/dd5cd6d782bf878d3c33ad6ed0df05c673dd61f6..48eb2e6ace25e9c96a0b468a65a0c228e8494790), and run it all by Joy. Could you please take another look? 

One section (`NewGalaxyModal.renderCreateWarning()`) that I revamped was per Emily's and Joy's feedback. It used to look like 

![image](https://user-images.githubusercontent.com/5438223/113570419-481deb80-95e2-11eb-98b2-601c2e356751.png)

and now it looks like

![image](https://user-images.githubusercontent.com/5438223/113570362-2a508680-95e2-11eb-8a56-69d457d5034c.png)

jdcanas commented on src/components/RuntimeManager.js:
```diff
@@ -297,16 +298,16 @@ export default class RuntimeManager extends PureComponent {
           div([
             div({ style: { fontSize: 12, fontWeight: 'bold' } }, ['Galaxy']),
-            div({ style: { fontSize: 10, textTransform: 'uppercase' } }, [app.status])
+            div({ style: { fontSize: 10, textTransform: 'capitalize' } }, [_.capitalize(app.status)])
```
I believe the use of the `textTransform` property in this div is duplicative of calling `_.capitalize` on the contents.

kyuksel replied:
Hmm...interestingly, `textTransform: 'capitalize'` doesn't seem to have any effect regardless of `_.capitalize(...)` so I'm taking that out.

rtitle commented on src/components/NewGalaxyModal.js:
```diff
@@ -186,59 +228,119 @@ export const NewGalaxyModal = _.flow(
+  const renderPersistentDiskSection = () => {
+    const gridStyle = { display: 'grid', gridTemplateColumns: '0.75fr 4.5rem 1fr 5.5rem 1fr 5.5rem', gridGap: '1rem', alignItems: 'center' }
+    return div({ style: { ...styles.whiteBoxContainer, marginTop: '1rem' } }, [
+      div({ style: styles.headerText }, ['Persistent Disk']),
```
Perhaps this should be `Persistent disk` to match the capitalization scheme of the rest of the page.

![image](https://user-images.githubusercontent.com/5368863/113600725-6bf72680-960e-11eb-91f2-51f0a5da846f.png)

(I don't know why capitalization sticks out to me so much.. sorry if this is being too nit-picky 😁 )

kyuksel replied:
Oh yea totally - for some reason I saw `Persistent Disk` at the same level as `Cloud Environment` 🤦🏼 

It's refreshing to see I'm not the only one with stylistic nits 😊

rtitle reviewed (approved):
I tested quite a few cases locally (creating apps with non-standard configs, pause, resume, delete, etc). Everything looks good (modulo the bug https://github.com/DataBiosphere/leonardo/pull/1996 fixes). The code diff also looks good to me (capitalization nits aside ;)).

One general comment: as discussed above, it also felt weird to me to be able to change the compute settings for a `Running` app, with no way to actually update it besides delete and recreate. Perhaps the boxes could be made uneditable for `Running` apps (though they should still be displayed). I think the warning at app creation time covers this though, so I don't feel that strongly to make that change in this PR.

Nice work!

xuscrbw reviewed (approved):
I think things look good. 

One difference I did notice was when the app is paused we have hover text for the "Delete" button in the Galaxy Modal, but we don't have similar hover text when the app is pausing or resuming. Is that intentional?

kyuksel commented:
> One difference I did notice was when the app is paused we have hover text for the "Delete" button in the Galaxy Modal, but we don't have similar hover text when the app is pausing or resuming. Is that intentional?

@xuscrbw Good catch! I've added a slightly modified version of the hover text to the default case so it will be displayed for the missing cases.

kyuksel committed:
PR feedback

kyuksel committed:
Merge remote-tracking branch 'origin/dev' into ky_ia_2389_custom_nodepool_config

kyuksel committed:
Merge remote-tracking branch 'origin/dev' into ky_ia_2389_custom_nodepool_config
